test/session-pool.js

```diff
@@ -16,2022 +16,1575 @@
+          assert.strictEqual(sessionPool.requestQueue_._concurrency, 11);
+        });
+      });
     });
-  });
 
-  describe('formatTrace_', function() {
-    var fakeFileName = 'path/to/file.js';
-    var fakeLineNumber = '99';
-    var fakeColumnNumber = '13';
-    var file = `${fakeFileName}:${fakeLineNumber}:${fakeColumnNumber}`;
+    describe('session_', function() {
+      it('should return a session object', function() {
+        const fakeSession = {};
 
-    var fakeFunction;
-    var fakeMethod;
+        DATABASE.session_ = function() {
+          return fakeSession;
+        };
 
-    var fakeTrace = [
-      {},
-      {},
-      {
-        getFunctionName: function() {
-          return fakeFunction;
-        },
-        getMethodName: function() {
-          return fakeMethod;
-        },
-        getFileName: function() {
-          return fakeFileName;
-        },
-        getLineNumber: function() {
-          return fakeLineNumber;
-        },
-        getColumnNumber: function() {
-          return fakeColumnNumber;
-        },
-      },
-    ];
+        const session = sessionPool.session_();
 
-    it('should return a trace with the method name', function() {
-      fakeMethod = 'MyClass.myMethod';
+        assert.strictEqual(session, fakeSession);
+        assert(isAround(session.lastUsed, Date.now()));
+      });
+    });
 
-      var expected = `Session leak detected!\n    at ${fakeMethod} (${file})`;
-      var actual = SessionPool.formatTrace_(fakeTrace);
+    describe('createReadSession', function() {
+      it('should create a read only session', function() {
+        let created = false;
 
-      assert.strictEqual(expected, actual);
-    });
+        const fakeSession = {
+          create: function() {
+            created = true;
+            return Promise.resolve();
+          },
+        };
+        sessionPool.session_ = function() {
+          return fakeSession;
+        };
+        return sessionPool.createReadSession().then(function(session) {
+          assert.strictEqual(session, fakeSession);
+          assert.strictEqual(session.type, 'readonly');
+          assert.strictEqual(created, true);
+        });
+      });
 
-    it('should return a trace with the function name', function() {
-      fakeFunction = 'myFunction';
+      it('should fail in creating a read only session', function() {
+        let deleteCalled = false;
+        const fakeSession = {
+          create: () => Promise.reject('Error'),
+          delete: () => {
+            deleteCalled = true;
+            return Promise.resolve();
+          },
+        };
 
-      var expected = `Session leak detected!\n    at ${fakeFunction} (${file})`;
-      var actual = SessionPool.formatTrace_(fakeTrace);
+        sessionPool.session_ = function() {
+          return fakeSession;
+        };
 
-      assert.strictEqual(expected, actual);
+        return sessionPool.createReadSession().catch(error => {
```

Comment:
We need to handle the non-failure case here, otherwise the test can potentially give a false positive.

```js
return sessionPool.createReadSession(
  () => { throw new Error('Should not pass.') },
  err => { ... }
);
```

Reply:
fixed it